Fix metadata override in Database Replicated due to RENAME after ALTER

evillique opened the pull request:
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix metadata override in Database Replicated which could have happened due to a RENAME query executed right after an ALTER query.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->

evillique committed:
Fix metadata override due to RENAME after ALTER

azat commented:
>[Integration tests (asan, flaky check)](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=78107&sha=5074002082de30bc46391d0bf3114855d02e068b&name_0=PR&name_1=Integration%20tests%20%28asan%2C%20flaky%20check%29) — Failed tests found: error: 0, passed: 93, fail: 3, skipped: 0, broken: 0, not_fail: 0

```
E           helpers.client.QueryRuntimeException: Client failed! Return code: 122, stderr: Received exception from server (version 25.4.1):
E           Code: 122. DB::Exception: Received from 172.16.1.9:9000. DB::Exception: Table columns structure in ZooKeeper is different from local table structure. Local columns:
E           columns format version: 1
E           1 columns:
E           `n` Int32
E           
E           Zookeeper columns:
E           columns format version: 1
E           1 columns:
E           `n` String
```

Let's retry the error in test?

evillique committed:
Help debug tests

evillique commented:
### test_replicated_database/test.py::test_force_synchronous_settings[3-3]
The error in this test was caused by the table `test_force_synchronous_settings.t` not being correctly removed in the previous run. The database `test_force_synchronous_settings` on the `dummy_replica` ran into this:
```
2025.03.22 05:42:53.265796 [ 16645 ] {} <Information> DatabaseReplicated (test_force_synchronous_settings): Executing CREATE TABLE test_force_synchronous_settings.t UUID '05b3385a-7d04-4563-a88a-d9b5dd76ec2c' (`n` String) ENGINE = ReplicatedMergeTree('/test/same/path/{shard}', '{replica}') ORDER BY tuple() SETTINGS index_granularity = 8192
2025.03.22 05:42:53.273127 [ 16645 ] {} <Debug> test_force_synchronous_settings.t (05b3385a-7d04-4563-a88a-d9b5dd76ec2c): Loading data parts
2025.03.22 05:42:53.273344 [ 16645 ] {} <Test> test_force_synchronous_settings.t (05b3385a-7d04-4563-a88a-d9b5dd76ec2c): loadDataParts: clearing data_parts_indexes (had 0 parts)
2025.03.22 05:42:53.273387 [ 16645 ] {} <Debug> test_force_synchronous_settings.t (05b3385a-7d04-4563-a88a-d9b5dd76ec2c): There are no data parts
2025.03.22 05:42:53.278704 [ 16645 ] {} <Debug> test_force_synchronous_settings.t (05b3385a-7d04-4563-a88a-d9b5dd76ec2c): This table /test/same/path/1 is already created, will add new replica
2025.03.22 05:42:53.281638 [ 16645 ] {} <Debug> test_force_synchronous_settings.t (05b3385a-7d04-4563-a88a-d9b5dd76ec2c): Creating replica /test/same/path/1/replicas/2
2025.03.22 05:42:53.330719 [ 16683 ] {} <Error> DDLWorker(test_force_synchronous_settings): Error on initialization of test_force_synchronous_settings: Code: 253. DB::Exception: Replica /test/same/path/1/replicas/2 already exists. (REPLICA_ALREADY_EXISTS), Stack trace (when copying this message, always include the lines below):
```
So when the table was removed it did not remove this last replica, therefore the table data in ZK, and after that when we tried to create a table with the same path but with different table structure it threw the exception in question.

It is not related to the current changes, and it is impossible to debug since related logs on the `dummy_node` were rotated and removed.

### test_replicated_database/test.py::test_recover_staled_replica[3-3]
Unclear what happened, fixed logging to shine the light on that

evillique committed:
Merge remote-tracking branch 'upstream/master' into fix-metadata-override

evillique commented:
[Stress test (msan)](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=78107&sha=fc881dbb83ebee60b930f204b06a6034347176d5&name_0=PR&name_1=Stress%20test%20%28msan%29) failure looks suspicious:
```
 [ 3328 ] {6a9e2010-e9dd-4078-93af-b7d54df429de} <Fatal> : Logical error: \'Digest does not match\'.
```

azat commented:
>[ 3328 ] {6a9e2010-e9dd-4078-93af-b7d54df429de} <Fatal> : Logical error: \'Digest does not match\'.

I did not verified this one particular, but likely it is not related (see https://github.com/ClickHouse/ClickHouse/issues/64936 for details), besides the query in crash is DROP

evillique committed:
Fix
